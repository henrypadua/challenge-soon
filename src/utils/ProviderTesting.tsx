import { NativeBaseProvider, NativeBaseProviderProps } from 'native-base';

import { theme } from '@styles/theme';

export const ProviderTesting = (props: NativeBaseProviderProps) => {
	const inset = {
		frame: { x: 0, y: 0, width: 0, height: 0 },
		insets: { top: 0, left: 0, right: 0, bottom: 0 },
	};

	return (
		<NativeBaseProvider initialWindowMetrics={inset} theme={theme}>
			{props.children}
		</NativeBaseProvider>
	);
};
