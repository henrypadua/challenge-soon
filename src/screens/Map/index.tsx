import React, { useState, useRef } from 'react';
import { Dimensions } from 'react-native';

import { GOOGLE_MAPS_APIKEY } from '@env';
import { useMap } from '@hooks/useMap';
import { Button, Image, Spinner, Text, VStack } from 'native-base';
import MapView, { Marker } from 'react-native-maps';
import MapViewDirections from 'react-native-maps-directions';

import utils from '@utils/Utils';

import { ModalInfo } from '@components/ModalInfo';

import { mapStyle } from '@styles/mapStyle';

export function Map() {
	const [isReady, setIsReady] = useState(false);
	const [angle, setAngle] = useState(0);
	const [infoRoute, setInfoRoute] = useState({
		distance: 0,
		duration: 0,
	});
	const [openModal, setOpenModal] = useState(false);

	const mapView = useRef<MapView>(null);
	const { isLoading, origin, destination } = useMap();

	return (
		<>
			{isLoading ? (
				<VStack
					flex={1}
					justifyContent="center"
					alignItems="center"
					bgColor={'#242E42'}
				>
					<Spinner size={'lg'} color="#ffffff" />
					<Text color="#ffffff">Loading...</Text>
				</VStack>
			) : (
				<MapView
					ref={mapView}
					style={{
						flex: 1,
						width: Dimensions.get('window').width,
						height: Dimensions.get('window').height,
					}}
					customMapStyle={mapStyle}
					region={{
						latitude: origin?.latitude,
						longitude: origin?.longitude,
						latitudeDelta: 0.0922,
						longitudeDelta: 0.0421,
					}}
					loadingBackgroundColor="#242E42"
					loadingIndicatorColor="#ffffff"
					loadingEnabled
					onMapLoaded={() => {
						setIsReady(true);
						setOpenModal(true);
					}}
					showsBuildings={false}
					showsPointsOfInterest={false}
				>
					{origin && (
						<Marker
							coordinate={origin}
							tracksViewChanges={true}
							anchor={{
								x: 0.5,
								y: 0.5,
							}}
						>
							<Image
								source={require('@assets/pin.png')}
								height={'50px'}
								width={'32px'}
								alt="pin"
							/>
						</Marker>
					)}

					{destination && (
						<Marker
							coordinate={destination}
							tracksViewChanges={true}
							anchor={{
								x: 0.5,
								y: 0.5,
							}}
							rotation={angle}
						>
							<Image
								source={require('@assets/car.png')}
								height={'50px'}
								width={'32px'}
								alt="pin"
							/>
						</Marker>
					)}
					<MapViewDirections
						origin={origin}
						destination={destination}
						apikey={GOOGLE_MAPS_APIKEY}
						strokeWidth={5}
						strokeColor="#9722FB"
						optimizeWaypoints={true}
						onReady={(result) => {
							setInfoRoute({
								distance: result.distance,
								duration: result.duration,
							});

							if (!isReady) {
								// Fit the map based on the route
								mapView?.current?.fitToCoordinates(result.coordinates, {
									edgePadding: {
										right: Dimensions.get('window').width * 0.1,
										bottom: 400,
										left: Dimensions.get('window').width * 0.1,
										top: Dimensions.get('window').height * 0.1,
									},
								});
								// Reposition the navigator
								if (result.coordinates.length >= 2) {
									const calculatedAngle = utils.calculateAngle(
										result.coordinates,
									);
									setAngle(calculatedAngle);
								}
								setIsReady(true);
							}
						}}
					/>
				</MapView>
			)}

			<ModalInfo
				infoRoute={infoRoute}
				openModal={openModal}
				setOpenModal={setOpenModal}
			/>
		</>
	);
}
