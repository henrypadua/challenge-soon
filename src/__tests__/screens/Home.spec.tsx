import { render } from '@testing-library/react-native';

import { ProviderTesting } from '@utils/ProviderTesting';

import { Home } from '@screens/Map';

jest.useFakeTimers();

test('renders correctly', () => {
	const screen = render(
		<ProviderTesting>
			<Home />
		</ProviderTesting>,
	).toJSON();

	expect(screen).toMatchSnapshot();
});
