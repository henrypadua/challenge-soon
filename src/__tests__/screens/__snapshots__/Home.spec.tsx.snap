// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
[
  <View
    dataSet={{}}
    style={
      {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "height": "100%",
      }
    }
  >
    <View
      pointerEvents="box-none"
      style={
        [
          {},
          {
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ]
      }
    >
      <View
        style={
          [
            {},
            undefined,
          ]
        }
      >
        <TextInput
          clearButtonMode="while-editing"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Onde você está?"
          style={
            [
              {},
              {
                "backgroundColor": "#eee",
                "borderRadius": 10,
                "fontSize": 15,
                "marginLeft": 40,
                "marginRight": 15,
                "marginVertical": 5,
                "padding": 10,
              },
            ]
          }
          value=""
        />
      </View>
    </View>
    <View
      dataSet={{}}
      style={{}}
    />
    <View
      pointerEvents="box-none"
      style={
        [
          {},
          {
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 55,
          },
        ]
      }
    >
      <View
        style={
          [
            {},
            undefined,
          ]
        }
      >
        <TextInput
          clearButtonMode="while-editing"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Para onde vamos?"
          style={
            [
              {},
              {
                "backgroundColor": "#eee",
                "borderRadius": 10,
                "fontSize": 15,
                "marginLeft": 40,
                "marginRight": 15,
                "marginVertical": 5,
                "padding": 10,
              },
            ]
          }
          value=""
        />
      </View>
    </View>
    <View
      dataSet={{}}
      style={{}}
    />
    <View
      dataSet={{}}
      style={
        {
          "flex": 1,
          "position": "absolute",
        }
      }
    >
      <Image
        accessibilityLabel="circle"
        alt="circle"
        dataSet={{}}
        onError={[Function]}
        source={1}
        style={
          [
            {
              "maxWidth": "100%",
            },
            {
              "left": 10,
              "position": "absolute",
              "top": 20,
            },
          ]
        }
      />
      <View
        dataSet={{}}
        style={{}}
      />
      <Image
        accessibilityLabel="line"
        alt="line"
        dataSet={{}}
        onError={[Function]}
        source={1}
        style={
          [
            {
              "maxWidth": "100%",
            },
            {
              "left": 18,
              "position": "absolute",
              "top": 40,
            },
          ]
        }
      />
      <View
        dataSet={{}}
        style={{}}
      />
      <Image
        accessibilityLabel="square"
        alt="square"
        dataSet={{}}
        onError={[Function]}
        source={1}
        style={
          [
            {
              "maxWidth": "100%",
            },
            {
              "left": 10,
              "position": "absolute",
              "top": 75,
            },
          ]
        }
      />
    </View>
  </View>,
  <View
    dataSet={{}}
    style={
      {
        "paddingBottom": 20,
        "paddingLeft": 40,
        "paddingRight": 40,
        "paddingTop": 20,
        "zIndex": 1,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      colorScheme="primary"
      dataSet={{}}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#242E42",
          "borderRadius": 4,
          "color": "#FFFFFF",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 10,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 10,
        }
      }
    >
      <View
        dataSet={{}}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
        test={true}
      >
        <View
          dataSet={{}}
          style={{}}
        >
          <Text
            dataSet={{}}
            style={
              {
                "backgroundColor": undefined,
                "color": "#FFFFFF",
                "fontFamily": "Inter_500Medium",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 21,
                "textDecorationLine": undefined,
              }
            }
          >
            Search Hero
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
